### ISSUE TITLE
Meta: Make distributed deployments possible

### ISSUE THREAD

**Issue by foxcpp** [foxcpp/maddy#279]
In the light of #277 and Project Hexanet's internal needs. Here is the list of things to do:

- [x] #259

Refactoring:
- [x] #303 
- [x] Replace updatepipe package written for go-imap v1 with a better update dispatcher.
- [ ] Implement support for some distributed broker - to be used with IMAP updates distribution.
- [ ] #262.
- [x] #304 

Several queries should be optimized/reworked in go-imap-sql in order to behave well in a distributed environment:
- [ ] https://github.com/foxcpp/go-imap-sql/issues/28
- [ ] https://github.com/foxcpp/go-imap-sql/issues/33

Officially support CockroachDB:
- [x] https://github.com/foxcpp/go-imap-sql/issues/5
- [ ] Officially support CockroachDB use with auth.sql_table/auth.sql_query.

In other words, we have a long way to go. :)

**Comment by savchenko** [comment]
Is this relevant to https://github.com/foxcpp/maddy/discussions/473 by any chance?..

**Comment by foxcpp** [comment]
Yes, it is directly related. Currently, maddy does but support any deployment setup using active replication. For MX redundancy, your best solution is to configure second MX to buffer incoming messages and attempt resending them to the main server infinitely.

**Comment by aviallon** [comment]
> Yes, it is directly related. Currently, maddy does but support any deployment setup using active replication. For MX redundancy, your best solution is to configure second MX to buffer incoming messages and attempt resending them to the main server infinitely.

Is there any hidden documentation on how to configure go-imap-sql with maddy?

I plan to setup a test server, which I'll migrate to when I feel the project is mature enough (I'm tired of dovecot).

**Comment by foxcpp** [comment]
go-imap-sql is storage.imapsql module in maddy. All relevant documentation is available here: https://maddy.email/reference/storage/imapsql/
Some additional (mostly technical) information is available here: https://github.com/foxcpp/go-imap-sql

**Comment by foxcpp** [comment]
> I'll migrate to when I feel the project is mature enough

As of that, go-imap-sql will be eventually replaced by https://github.com/foxcpp/maddy-storage that is is in early development stage currently.
